### ISSUE TITLE
Track Web hook fail open

### ISSUE THREAD

**Issue by ltagliamonte-dd** [kubernetes/kubernetes#107146]
### What would you like to be added?

Add a prometheus metrics to track when a webhook (admission or mutating) fails open.

### Why is this needed?

In some situations cluster admin (or app/chart developers) prefers to fail open when a requests towards a mutation or admission web hook fails.
However is it important to monitor and alert on those conditions, currently the only way to do it is by alerting on api server logs:
```
Failed calling webhook, failing open
```

**Comment by ltagliamonte-dd** [comment]
@fedebongio i have a branch where i've added the metric, happy to submit a PR

**Comment by roycaihw** [comment]
To distinguish a [fail open](https://github.com/kubernetes/kubernetes/blob/38ec822b46853acfb3c3dc7e386a8d4636c67e03/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L153) from a [successful admit](https://github.com/kubernetes/kubernetes/blob/38ec822b46853acfb3c3dc7e386a8d4636c67e03/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L164), does it work for you if you filter `webhook_request_total` by `!rejected && code != 200`?

**Comment by ltagliamonte-dd** [comment]
Thanks for the suggestion, I'm on 1.21 and i see the metric `webhook_request_total` been added to 1.23.
However looking briefly at the code it doesn't seems the `ObserveWebhook` gets invoked on fail open.
My goal here is to precisely monitor requests that fail open.
If you don't want to add a dedicated metric if you want we can dedicate an error code to fail open and use `ObserveWebhook`
What do you think @roycaihw ?

**Comment by roycaihw** [comment]
> it doesn't seems the `ObserveWebhook` gets invoked on fail open

IIUC it gets invoked [here](https://github.com/kubernetes/kubernetes/blob/38ec822b46853acfb3c3dc7e386a8d4636c67e03/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L153)

**Comment by ltagliamonte-dd** [comment]
@roycaihw yes you are right, it looks like it is the only case in which `rejected` will be `false`.
Just wondering if a code refactoring would break this assumption in the future, ergo breaking alerting.

A dedicated metric looks a bit more appealing to me to someone that look at that piece of code in few years from now, but if there are reasons to don't add a dedicated metric, I can understand.

I don't have a 1.23 to test but it seem pretty straightforward.

**Comment by roycaihw** [comment]
I think it's okay to have a dedicated metric. Curious what's SIG instrumentation's recommendation.

/sig instrumentation

One place where the existing `webhook_request_total` may be insufficient is if an ErrCallingWebhook with code 200 gets returned, in which case we won't be able to distinguish it from a successfully admit. An alternative to adding a dedicated metric is to add a boolean "fail open" label to the `webhook_request_total` metric. However it could double the cardinality, which in theory is already large due to the HTTP code label (0~600).

/unassign
/assign @ltagliamonte-dd

**Comment by ltagliamonte-dd** [comment]
@roycaihw since this is my first contribution, i'd like to understand how to proceed with this?
are we waiting on sig/instrumentation?

**Comment by roycaihw** [comment]
I've lgtm'ed the PR. You will need an [owner's approval](https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md#the-code-review-process) for it to be merged.

**Comment by dashpole** [comment]
cc @logicalhan

**Comment by dashpole** [comment]
Can this be closed?
